java-checks/src/main/java/org/sonar/java/checks/regex/ImpossibleBoundariesCheck.java

```diff
@@ -69,26 +81,74 @@ public void visitLookAround(LookAroundTree tree) {
       }
     }
 
+    @Override
+    public void visitDisjunction(DisjunctionTree tree) {
+      BoundaryInDisjunctionFounder boundaryInDisjunctionFounder = new BoundaryInDisjunctionFounder();
+      boundaryInDisjunctionFounder.visit(tree);
+      excluded.addAll(boundaryInDisjunctionFounder.foundBoundaries());
+      super.visitDisjunction(tree);
+    }
+
     @Override
     public void visitBoundary(BoundaryTree boundaryTree) {
       switch (boundaryTree.type()) {
         case LINE_START:
         case INPUT_START:
           if (!RegexTreeHelper.canReachWithoutConsumingInput(start, boundaryTree)) {
             reportIssue(boundaryTree, String.format(MESSAGE, "after"), null, Collections.emptyList());
+          } else if (!excluded.contains(boundaryTree) && probablyShouldConsumeInput(start, boundaryTree)) {
+            reportIssue(boundaryTree, String.format(SOFT_MESSAGE, "after"), null, Collections.emptyList());
           }
           break;
         case LINE_END:
         case INPUT_END:
         case INPUT_END_FINAL_TERMINATOR:
           if (!RegexTreeHelper.canReachWithoutConsumingInput(boundaryTree, end)) {
             reportIssue(boundaryTree, String.format(MESSAGE, "before"), null, Collections.emptyList());
+          } else if (!excluded.contains(boundaryTree) && probablyShouldConsumeInput(boundaryTree, end)) {
+            reportIssue(boundaryTree, String.format(SOFT_MESSAGE, "before"), null, Collections.emptyList());
           }
           break;
         default:
           // Do nothing
       }
     }
+
+    private boolean probablyShouldConsumeInput(AutomatonState start, AutomatonState goal) {
+      return  canReachWithConsumingInput(start, goal, new HashSet<>());
+    }
   }
 
+  private static class BoundaryInDisjunctionFounder extends RegexBaseVisitor {
+    private final Set<BoundaryTree> foundBoundaries = new HashSet<>();
+
+    @Override
+    public void visitBoundary(BoundaryTree boundaryTree) {
+      foundBoundaries.add(boundaryTree);
+    }
+
+    public Set<BoundaryTree> foundBoundaries() {
+      return new HashSet<>(foundBoundaries);
+    }
+  }
+
+  private boolean canReachWithConsumingInput(AutomatonState start, AutomatonState goal, Set<AutomatonState> visited) {
+    if (start == goal 
+      || start instanceof LookAroundTree 
+      || start.continuation() instanceof LookAroundTree 
+      || visited.contains(start)) {
+      return false;
+    }
+
+    visited.add(start);
+
+    for (AutomatonState successor : start.successors()) {
+      AutomatonState.TransitionType transition = successor.incomingTransitionType();
+      if (((transition == CHARACTER) && regexReachabilityChecker.canReach(successor, goal))
+        || canReachWithConsumingInput(successor, goal, visited)) {
```

Comment:
This still leads to some unnecessary work when `canReach` returns false because it will still try `canReachWithConsumingInput` even though it can never return true in that case (which is also why you can't get full coverage here). You should change it so that `canReachWithConsumingInput` is only called when the edge is not a character edge.